### ISSUE TITLE
git.latest broken behavior in 2015.8.x on older Git

### ISSUE THREAD

**Issue by eykd** [saltstack/salt#31229]
We rely heavily on `git.latest` for deploying our web properties. After upgrading from 2014.7.x to 2015.8.x this month, I've begun witnessing very strange behavior on Ubuntu Precise with Git 1.7.9.5. 

Here are steps to reproduce the problem as I experienced it, starting with 2014.7.x and then upgrading to 2015.8.x:

### Locally:

Set up the VM on step-1 tag:

```
git clone https://gist.github.com/bd959cbabe91aa690e17.git weird-git
cd weird-git
git checkout step-1
vagrant up
vagrant ssh
```

### On the VM:

Check the state of the VM after the highstate.

```
cd /opt/stuff/django
git branch  # Should be `stable/1.4.x`:
#   master
# * stable/1.4.x
git log -1  # Should be `5864e1f8e91d31d914c27885240ea6e065b38fd4`
logout
```

### Locally:

Switch to step-2 tag:

```
git checkout step-2
vagrant ssh
```

### On the VM:

Upgrade Salt to `v2015.8.5` and run the highstate.

```
sudo sh /root/install_salt.sh -P git v2015.8.5
sudo salt-call state.highstate
cd /opt/stuff/django
git branch  # Is still `stable/1.4.x`:
git log -1  # Should be `57671dd0b91ed4def9c7a49c486944f4e98fed9d` (HEAD of stable/1.5.x)
```

Note the branches now:

```
#   master
#   origin/stable/1.5.x
# * stable/1.4.x
```

First of all, this is just weird, and breaks my expectations of common git conventions. When I check out a branch, I expect the local branch to be the name of the branch I checked out, not whatever branch I was on.

### Run the highstate again

Here's where it gets weird:

```
sudo salt-call state.highstate  # This will fail on the `django` state.
```

OK, so fix that manually:

```
git checkout 57671dd0b91ed4def9c7a49c486944f4e98fed9d
git branch -d stable/1.4.x
git branch -d origin/stable/1.5.x
sudo salt-call state.highstate  # This will fail again on the `django` state
# fatal: Cannot setup tracking information; starting point is not a branch.
```

OK, so I guess we need to use that new `branch` option on the state, right?

```
logout
```

### Locally:

```
git checkout step-3
vagrant ssh
```

### On the VM:

```
sudo salt-call state.highstate  # Success! Fistpump!
sudo salt-call state.highstate  # Failure! Whaaaa?!!
# fatal: A branch named 'origin/stable/1.5.x' already exists.
```

I've also been able to produce behavior where a branch deploys an old revision, and not the HEAD, when switching between branchs, but I'm not sure how to reproduce it.

**Comment by eykd** [comment]
Oh, I should also say, I've definitely encountered this bug on Ubuntu Precise, with Git 1.7.9.5. I've not tested extensively, but I haven't yet encountered the weirdness on Ubuntu Trusty, with Git 1.9.1.

**Comment by eykd** [comment]
And I've run through the reproduction steps with Git 1.9.1, and I can't reproduce the bug, so the problem seems limited to the older Git. It seems to have to do with the upstream remote manipulation that `git.latest` is performing, but that's about the best I can say at the moment.

**Comment by jfindlay** [comment]
@eykd, thanks for reporting.

**Comment by jfindlay** [comment]
@eykd, it seems that this may be a duplicate of #30999, for which a fix is proposed in #31245.  Is it possible for you to test the fix?

**Comment by eykd** [comment]
Figuring out this bug took a lot of time, and now I'm backed up. Once I get a bit more caught up, I'll take a look at it.

**Comment by jfindlay** [comment]
@eykd, thanks, take the time you need.

**Comment by terminalmage** [comment]
@jfindlay I don't think this is the same thing as #30999 at all, that issue dealt with a tag, and how Salt tries to unset the tracking branch when the rev is a tag. This is a branch.

I will investigate later today.

**Comment by eykd** [comment]
Well, I thought I had a workaround, namely setting a local branch name that has no equivalent upstream, but now I'm getting failures on subsequent deploys again because git is creating the weird `origin/<branchname>` local branches, and then failing when it tries to create them again.

**Comment by terminalmage** [comment]
@eykd I found and fixed the issue. The main problem was that we were using ``--track`` on older git when we needed to use ``--set-upstream``.

See https://github.com/saltstack/salt/pull/31373 for details. Once this is merged, I'd love if you could test using your vagrant setup with a git install from the head of the 2015.8 branch.

Thank you so much for your extremely detailed report. I wish all bug reports were so well-investigated! I'm really sorry my code caused this problem for you, though. :frowning:

**Comment by eykd** [comment]
Splendid! I have a crucial app that's still on precise, and I was forced to deploy it by hand via wooden planks rough-hewn with a stone axe. I'll be happy to test the fix when it's merged.

Bugs are inevitable, but bugfixes aren't, so thanks for the quick turnaround!

**Comment by terminalmage** [comment]
@eykd Feel free to test, my pull request was merged into the 2015.8 branch.

**Comment by eykd** [comment]
Confirmed, the latest 2015.8 branch works for me on Ubuntu precise w/ Git 1.7.9.5.

**Comment by terminalmage** [comment]
Awesome, thanks for confirming.
